File: internal/linker/linker.go
package linker

import (
	"bufio"
	"encoding/json"
	"log"
	"net/url"
	"os"
	"path/filepath"
	"strings"
	"unicode/utf8"

	"golang.org/x/net/html"
	"golang.org/x/net/html/atom"

	"github.com/usami/wikilinks-jp/pkg/types"
)

type Linker struct {
	Category      string
	Annotations   []*types.Annotation
	Pages         map[json.Number]*types.Page
	TitleToPageID map[string]json.Number
}

func NewLinker(c string) *Linker {
	return &Linker{
		Category:      c,
		Annotations:   make([]*types.Annotation, 0),
		Pages:         make(map[json.Number]*types.Page),
		TitleToPageID: make(map[string]json.Number),
	}
}

func (l *Linker) Load(a, d, t string) {
	log.Printf("linker[%s]: load annotaions", l.Category)
	l.loadAnnotations(a)
	log.Printf("linker[%s]: load pages", l.Category)
	l.loadPages(d)
	log.Printf("linker[%s]: load title to pageid mappings", l.Category)
	l.loadTitlePageIDs(t)
}

func (l *Linker) Run() {
	log.Printf("linker[%s]: check links", l.Category)
	l.checkLinks()
}

func (l *Linker) Output(filepath string) {
	log.Printf("linker[%s]: output analyzed results", l.Category)

	f, err := os.OpenFile(filepath, os.O_RDWR|os.O_CREATE, 0644)
	if err != nil {
		log.Fatal(err)
	}

	for _, an := range l.Annotations {
		if an.LinkPageID != "" {
			jsonStr, err := json.Marshal(an)
			if err != nil {
				log.Fatal(err)
			}
			s := string(jsonStr) + "\n"

			if _, err := f.WriteString(s); err != nil {
				log.Fatal(err)
			}
		}
	}

	if err := f.Close(); err != nil {
		log.Fatal(err)
	}
}

func (l *Linker) loadAnnotations(filepath string) {
	f, err := os.Open(filepath)
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()

	s := bufio.NewScanner(f)

	var aa []*types.Annotation

	for s.Scan() {
		var an types.Annotation
		if err := json.Unmarshal([]byte(s.Text()), &an); err != nil {
			log.Fatal(err)
		}
		aa = append(aa, &an)
	}

	l.Annotations = aa
}

func (l *Linker) loadPages(dirpath string) {
	files := listHTMLFiles(dirpath)

	amap := make(map[json.Number][]int)

	for _, an := range l.Annotations {
		lines := amap[an.PageID]
		if lines == nil {
			lines = make([]int, 0)
		}
		lines = append(lines, an.HTMLOffset.Start.LineID)
		amap[an.PageID] = lines
	}

	for _, file := range files {
		pid := types.ParsePageID(file)
		page := types.LoadPageFromHTML(file, pid, amap[pid])
		l.Pages[page.PageID] = page
	}
}

func (l *Linker) loadTitlePageIDs(filepath string) {
	f, err := os.Open(filepath)
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()

	s := bufio.NewScanner(f)

	for s.Scan() {
		var t types.TitlePageIDWithRedirect
		if err := json.Unmarshal([]byte(s.Text()), &t); err != nil {
			log.Fatal(err)
		}
		l.TitleToPageID[t.Title] = t.Resolve()
	}
}

func (l *Linker) checkLinks() {
	for _, an := range l.Annotations {
		p := l.Pages[an.PageID]
		li := an.HTMLOffset.Start.LineID

		doc, err := html.Parse(strings.NewReader(p.Lines[li]))
		if err != nil {
			log.Fatal(err)
		}

		var f func(*html.Node, int) int

		f = func(n *html.Node, offset int) int {
			switch n.Type {
			case html.ElementNode:
				if n.DataAtom == atom.A {
					for _, attr := range n.Attr {
						if isLinkToEntity(attr) {
							if matchesAnnotation(n.FirstChild, an, offset) {
								title := extractTitle(attr.Val)
								if pageID, ok := l.TitleToPageID[title]; ok {
									an.LinkPageID = pageID
								}
								break
							}
						}
					}
				}
			case html.TextNode:
				offset += utf8.RuneCountInString(n.Data)
			}

			for c := n.FirstChild; c != nil; c = c.NextSibling {
				offset = f(c, offset)
			}

			return offset
		}
		f(doc, 0)
	}
}

func listHTMLFiles(dirpath string) []string {
	if _, err := os.Stat(dirpath); os.IsNotExist(err) {
		log.Fatal(err)
	}

	var files []string

	err := filepath.Walk(dirpath, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if !info.IsDir() && filepath.Ext(path) == ".html" {
			files = append(files, path)
		}
		return nil
	})

	if err != nil {
		log.Fatal(err)
	}

	return files
}

func isLinkToEntity(a html.Attribute) bool {
	return a.Key == "href" && (strings.HasPrefix(a.Val, "/index.php/") || strings.HasPrefix(a.Val, "/a-sumida/wiki2019_1/index.php/"))
}

func extractTitle(s string) string {
	u, err := url.Parse(s)
	if err != nil {
		log.Fatal(err)
	}
	parts := strings.Split(u.Path, "/")
	return parts[len(parts)-1]
}

func matchesAnnotation(n *html.Node, a *types.Annotation, offset int) bool {
	if n == nil || n.Type != html.TextNode {
		return false
	}

	return a.TextOffset.Start.Offset == offset && a.TextOffset.Text == n.Data
}

File: python/src/linker/linker.py
import json
import ujson
import logging
import os
from typing import List, Dict
from urllib.parse import urlparse, unquote

from bs4 import BeautifulSoup
from bs4.element import Tag, NavigableString
from dacite import from_dict

from src.types.annotation import Annotation
from src.types.page import Page, parse_page_id, load_page_from_html
from src.types.title_pageid import TitlePageIDWithRedirect

logger = logging.getLogger(__name__)
logging.basicConfig(
    format="%(asctime)s - %(levelname)s - %(name)s - %(message)s",
    datefmt="%m/%d/%Y %H:%M:%S",
    level=logging.INFO,
)


class Linker:
    def __init__(
            self, category: str, annotations: List[Annotation] = None,
            pages: Dict[str, Page] = None, title_to_page_id: Dict[str, int] = None) -> None:
        self.category: str = category
        self.annotations: List[Annotation] = [
        ] if annotations is None else annotations
        self.pages: Dict[str, Page] = {} if pages is None else pages
        self.title_to_page_id: Dict[str, int] = {
        } if title_to_page_id is None else title_to_page_id

    def load(self, a: str, d: str, t: str) -> None:
        logger.info("linker[%s]: load annotaions", self.category)
        self.load_annotations(a)
        logger.info("linker[%s]: load pages", self.category)
        self.load_pages(d)
        logger.info("linker[%s]: load title to pageid mappings", self.category)
        self.load_title_page_ids(t)

    def run(self) -> None:
        logger.info("linker[%s]: check links", self.category)
        self.check_links()

    def output(self, filepath: str) -> None:
        logger.info("linker[%s]: output analyzed results", self.category)

        with open(filepath, mode="w") as f:
            for an in self.annotations:
                if an.link_page_id != "":
                    json_str = an.to_json(ensure_ascii=False)
                    s = json_str + "\n"
                    f.write(s)

    def load_annotations(self, filepath: str) -> None:
        aa: List[Annotation] = []
        with open(filepath, mode="r") as f:
            for line in f:
                d = json.loads(line)
                an = from_dict(Annotation, d)
                aa.append(an)

        self.annotations = aa

    def load_pages(self, dirpath: str) -> None:
        files = list_html_files(dirpath)

        amap: Dict[str, List[int]] = {}

        for an in self.annotations:
            lines = amap.get(an.page_id, [])
            lines.append(an.html_offset.start.line_id)
            amap[an.page_id] = lines

        for f in files:
            pid = parse_page_id(f)
            if pid in amap:
                page = load_page_from_html(f, pid, amap[pid])
                self.pages[page.page_id] = page

    def load_title_page_ids(self, filepath: str) -> None:
        base_name = os.path.basename(filepath)
        dirname = os.path.dirname(filepath)
        dic_path = os.path.join(dirname, "dic-" + base_name)
        if os.path.exists(dic_path):
            logger.info(
                "exists dictionary file of title to page id. loading dictionary file...")
            with open(dic_path, mode="r") as f:
                self.title_to_page_id = json.load(f)
            return

        logger.info("not exists dictionary file of title to page id.")
        with open(filepath, mode="r") as f:
            for i, line in enumerate(f):
                if i % 10000 == 0:
                    logger.info("step of loading title page ids: %s", i)
                d = ujson.loads(line)
                if "redirect_to" in d:
                    if d["redirect_to"]["page_id"] is None or d["redirect_to"]["title"] is None:
                        del d["redirect_to"]
                        d["is_redirect"] = False

                t = from_dict(TitlePageIDWithRedirect, d)
                self.title_to_page_id[t.title] = t.resolve()

        with open(dic_path, mode="w") as f:
            json.dump(self.title_to_page_id, f, indent=2, ensure_ascii=False)

    def check_links(self):
        for an in self.annotations:
            p = self.pages[an.page_id]
            li = an.html_offset.start.line_id
            soup = BeautifulSoup(p.lines[li], 'html.parser')

            def f(n, offset: int) -> int:
                if type(n) == Tag:
                    if n.name == "a":
                        href = n.attrs["href"]
                        if is_link_to_entity(href):
                            if hasattr(n, "children"):
                                first_child = list(n.children)[0]
                                if matches_annotation(first_child, an, offset):
                                    title = extract_title(href)
                                    if title in self.title_to_page_id:
                                        page_id = self.title_to_page_id[title]
                                        an.link_page_id = str(page_id)
                elif type(n) == NavigableString:
                    offset += len(n.string)

                if hasattr(n, "children"):
                    for c in n.children:
                        offset = f(c, offset)

                return offset

            f(soup, 0)


def list_html_files(dirpath: str) -> List[str]:
    file_list: List[str] = []
    for cur_dir, dirs, files in os.walk(dirpath):
        for f in files:
            _, ext = os.path.splitext(f)
            if ext == '.html':
                path = os.path.join(cur_dir, f)
                file_list.append(path)
    return file_list


def is_link_to_entity(a: str) -> bool:
    return a.startswith("/index.php/") or a.startswith("/a-sumida/wiki2019_1/index.php/")


def extract_title(s: str) -> str:
    u = urlparse(s)
    path = u.path
    parts = path.split("/")
    title = unquote(parts[len(parts)-1])
    return title


def matches_annotation(n, a: Annotation, offset: int) -> bool:
    if not type(n) == NavigableString:
        return False
    return a.text_offset.start.offset == offset and a.text_offset.text == n.string
